keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1502 @@
+
+## Proposal
+
+### Summary
+
+Before getting into all the individual fields and capabilities, let's look at the general "shape" of the API.
+Very similar to what we have in ValidatingAdmissionPolicy, this API separates policy definition from policy configuration by splitting responsibilities across resources. The resources involved are:
+- Policy definitions (MutatingAdmissionPolicy)
+- Policy bindings (MutatingAdmissionPolicyBinding)
+- Policy param resources (custom resources or config maps)
+
+The idea is to leverage the CEL power of the object construction and allow users to define how they want to mutate the admission request through CEL expression.
+This proposal aims to allow mutations to be expressed using the "apply configuration" introduced by Server Side Apply. 
+And users would be able to define only the fields they care about inside MutatingAdmissionPolicy, the object will be constructed using CEL which would be similar to a Server Side Apply configuration patch and then be merged into the request object using the structural merge strategy. 
+See sigs.k8s.io/structured-merge-diff for more details.
+
+Note: See the alternative consideration section for the alternatives.
+
+Pros:
+- Consistent with Server Side Apply so that we will continue investing SSA as the best way to do patch updates to resources;
+- Does not require the users to learn a new syntax;
+- Inherit the declarative nature;
+- Existing merging strategy to be leverage without rewritten and the effort of maintaining multiple systems;
+- Support the existing makers/openapi extensions.
+
+Cons:
+- Lack of deletion support (see the unsetting values section for the details and workaround);
+- Migration effort from Mutation Webhook
+  
+
+### Phase 1
+
+#### API Shape
+Similar to the `validations` field in ValidatingAdmissionPolicy, a `mutations` field will be defined inside MutatingAdmissionPolicy which allows users to define a list of mutations that apply to the specific resources.
+Each mutation field contains a CEL expression which evaluates to a partially populated object representing a Server Side Apply "apply configuration". The apply configuration is then merged into the request object.
+
+Here is an example of injecting an initContainer.
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: MutatingAdmissionPolicy
+metadata:
+  name: "sidecar-policy.example.com"
+spec:
+  paramKind:
+    group: mutations.example.com
+    kind: Sidecar
+    version: v1
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["apps"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE"]
+      resources:   ["pods"]
+  matchConditions:
+    - name: does-not-already-have-sidecar
+      expression: "!object.spec.initContainers.exists(ic, ic.name == params.name)"
+  failurePolicy: Fail
+  reinvocationPolicy: IfNeeded
+  mutations:
+    - patchType: "ApplyConfiguration" // "ApplyConfiguration", "JSONPatch" supported. Default to "ApplyConfiguration"
```

Comment:
no default on union discriminators, but I think this shape gives us the flexibility we'll need.